executor/src/witgen/jit/affine_symbolic_expression.rs

```diff
@@ -0,0 +1,550 @@
+            return vec![];
+        }
+
+        // We need to assert that the masks cover the offset,
+        // otherwise the equation is not solvable.
+        // We assert offset & !masks == 0 <=> offset == offset | masks.
+        // We use the latter since we cannot properly bit-negate inside the field.
+        effects.push(Assertion::assert_eq(
+            self.offset.clone(),
+            &self.offset | &T::from(covered_bits).into(),
+        ));
+
+        effects
+    }
+
+    fn transfer_constraints(
+        &self,
+        range_constraints: &impl RangeConstraintSet<V, T>,
+    ) -> Vec<Effect<T, V>> {
+        // We are looking for X = a * Y + b * Z + ... or -X = a * Y + b * Z + ...
+        // where X is least constrained.
+
+        let Some((solve_for, solve_for_coefficient)) = self
+            .coefficients
+            .iter()
+            .filter(|(_var, coeff)| coeff.is_known_one() || coeff.is_known_minus_one())
+            .max_by_key(|(var, _c)| {
+                // Sort so that we get the least constrained variable.
+                range_constraints
+                    .range_constraint((*var).clone())
+                    .map(|c| c.range_width())
+                    .unwrap_or_else(|| T::modulus())
+            })
+        else {
+            return vec![];
+        };
+
+        // This only works if the coefficients are all known.
+        let Some(summands) = self
+            .coefficients
+            .iter()
+            .filter(|(var, _)| *var != solve_for)
+            .map(|(var, coeff)| {
+                coeff.try_to_number().and_then(|coeff| {
+                    range_constraints
+                        .range_constraint(var.clone())
+                        .map(|constraint| constraint.multiple(coeff))
+                })
+            })
+            .chain(std::iter::once(self.offset.range_constraint()))
+            .collect::<Option<Vec<_>>>()
+        else {
+            return vec![];
+        };
+        let Some(constraint) = summands.into_iter().reduce(|c1, c2| c1.combine_sum(&c2)) else {
+            return vec![];
+        };
+        let constraint = if solve_for_coefficient.is_known_one() {
+            -constraint
+        } else {
+            constraint
+        };
+        vec![Effect::RangeConstraint(solve_for.clone(), constraint)]
+    }
+}
+
+impl<T: FieldElement, V: Clone + Ord> Add for &AffineSymbolicExpression<T, V> {
+    type Output = AffineSymbolicExpression<T, V>;
+
+    fn add(self, rhs: Self) -> Self::Output {
+        let mut coefficients = self.coefficients.clone();
+        for (var, coeff) in rhs.coefficients.iter() {
+            coefficients
+                .entry(var.clone())
+                .and_modify(|f| *f = &*f + coeff)
+                .or_insert_with(|| coeff.clone());
+        }
+        let coefficients = coefficients
+            .into_iter()
+            .filter(|(_, f)| !f.is_known_zero())
+            .collect();
```

Comment:
Can use [`retain()`](https://doc.rust-lang.org/std/collections/struct.BTreeMap.html#method.retain) instead.

But maybe it would be even better to use [`remove_entry()`](https://doc.rust-lang.org/std/collections/btree_map/struct.OccupiedEntry.html#method.remove_entry) on the previous loop, but probably wouldn't look so nice and idiomatic.